stl/inc/flat_map

```diff
@@ -960,210 +886,526 @@ private:
 flat_map(sorted_unique_t, _KeyContainer, _MappedContainer, _Compare, _Allocator)
     -> flat_map<typename _KeyContainer::value_type, typename _MappedContainer::value_type, _Compare, _KeyContainer,
         _MappedContainer>;
 
-template <class _InputIterator, class _Compare = less<typename iter_value_t<_InputIterator>::first_type>>
+template <class _InputIterator, class _Compare = less<_Guide_key_t<_InputIterator>>>
 flat_map(_InputIterator, _InputIterator, _Compare = _Compare())
-    -> flat_map<typename iter_value_t<_InputIterator>::first_type, typename iter_value_t<_InputIterator>::second_type,
-        _Compare>;
+    -> flat_map<_Guide_key_t<_InputIterator>, _Guide_val_t<_InputIterator>, _Compare>;
 
-template <class _InputIterator, class _Compare = less<typename iter_value_t<_InputIterator>::first_type>>
+template <class _InputIterator, class _Compare = less<_Guide_key_t<_InputIterator>>>
 flat_map(sorted_unique_t, _InputIterator, _InputIterator, _Compare = _Compare())
-    -> flat_map<typename iter_value_t<_InputIterator>::first_type, typename iter_value_t<_InputIterator>::second_type,
-        _Compare>;
-
-template <_RANGES input_range _R, class _Compare = less<typename _RANGES range_value_t<_R>::first_type>,
-    class _Allocator = allocator<byte>>
-flat_map(from_range_t, _R&&, _Compare = _Compare(), _Allocator = _Allocator())
-    -> flat_map<typename _RANGES range_value_t<_R>::first_type, typename _RANGES range_value_t<_R>::second_type,
-        _Compare,
-        vector<typename _RANGES range_value_t<_R>::first_type,
-            _Rebind_alloc_t<_Allocator, typename _RANGES range_value_t<_R>::first_type>>,
-        vector<typename _RANGES range_value_t<_R>::second_type,
-            _Rebind_alloc_t<_Allocator, typename _RANGES range_value_t<_R>::second_type>>>;
-
-template <_RANGES input_range _R, class _Allocator>
-flat_map(from_range_t, _R&&, _Allocator) -> flat_map<typename _RANGES range_value_t<_R>::first_type,
-    typename _RANGES range_value_t<_R>::second_type, less<typename _RANGES range_value_t<_R>::first_type>,
-    vector<typename _RANGES range_value_t<_R>::first_type,
-        _Rebind_alloc_t<_Allocator, typename _RANGES range_value_t<_R>::first_type>>,
-    vector<typename _RANGES range_value_t<_R>::second_type,
-        _Rebind_alloc_t<_Allocator, typename _RANGES range_value_t<_R>::second_type>>>;
-
-template <class _Key, class _T, class _Compare = less<_Key>>
-flat_map(initializer_list<pair<_Key, _T>>, _Compare = _Compare()) -> flat_map<_Key, _T, _Compare>;
-
-template <class _Key, class _T, class _Compare = less<_Key>>
-flat_map(sorted_unique_t, initializer_list<pair<_Key, _T>>, _Compare = _Compare()) -> flat_map<_Key, _T, _Compare>;
+    -> flat_map<_Guide_key_t<_InputIterator>, _Guide_val_t<_InputIterator>, _Compare>;
+
+template <_RANGES input_range _Rng, class _Compare = less<_Range_key_type<_Rng>>, class _Allocator = allocator<byte>>

```

Comment:
The standard says `byte` here. However, it seems that we support opting-out `std::byte` by defining `_HAS_STD_BYTE` as `0`.
Should we switch to use `allocator<unsigned char>` (or something similar) in that case or just make the deduction guides unavailable?